### DISCUSSION TITLE
Code signing OSS casks for Apple Sillicon

### DISCUSSION THREAD

**Discussion by agologan** [Homebrew/discussions#1790] [Writing Formulae/Casks] [answered]
Hey, I maintain https://github.com/liquidx/webviewscreensaver I'm looking at any options/suggestions on how to approach the M1 signature requirement.
One idea could be to enhance brew with a code-signing stanza to invoke the `apply_ad_hoc_signature` after install.
Another is to migrate from cask to a formula though non-CLI artifacts are not usually encouraged in core.

Any other ideas? How have other cask maintainers approached this?

**Comment by carlocab** [comment, marked as answer]
[`apply_ad_hoc_signature`](https://github.com/Homebrew/brew/blob/10edae9b94d8dad41f04849a5f32d86256e23f23/Library/Homebrew/os/mac/keg.rb#L64) just calls `codesign`. This is only needed when using `install_name_tool` to change binary install names or rpaths. If your build doesn't do anything like that, there should be no need to apply an ad-hoc signature. If you do need to apply one, why don't you do so as part of your build?

**Reply by agologan** [reply]
The binary is distributed as an **unsigned** cask https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/webviewscreensaver.rb 
To run on Apple Sillicon it requires a signature. 
As far as I know ad hoc signing works only for local builds, not distributed binaries.
The *proper* way to do it is to pay for a developer account and then sign and notarize the binary using the Developer ID.

Since this is an open-source hobby project going that route does not seem reasonable.
So I'm looking for other recommendations from the community.

**Reply by SMillerDev** [reply]
Homebrew doesn't really have a solution for this. You'd be better off asking on the Apple forums or Stackoverflow.

**Reply by agologan** [reply]
After testing adhoc codesigning (`codesign --sign -`) and distribution at @carlocab suggestion it seems to work fine.

**Reply by tlk** [reply]
@agologan Thanks for the update.

I saw your change at https://github.com/liquidx/webviewscreensaver/commit/5cb16aa81d6b8687dc73f287afebff7201252649 and noticed that code signing is disabled manually when running `xcodebuild` with `CODE_SIGNING_REQUIRED=NO`, so perhaps not doing that would be an alternative way to solve the issue. Anyway, happy to hear that you found a way.

**Reply by agologan** [reply]
Adhoc codesigning is different then typical signatures because they don't contain an actual cryptographic message instead just adding sha of the object without any way of verifying the signing identity.
This seems to be enough to run the code on Apple Silicon but it doesn't get rid of the unidentified developer prompt and it doesn't enable notarisation.

To adhoc sign a build locally you can do `codesign --sign - <target>` on an unsigned object (or pass `--force` to resign) or during the build process by passing `CODE_SIGN_IDENTITY="-" CODE_SIGNING_REQUIRED=YES`.
I don't have a preference for either but it may be easier to specify that as part of the build if you have multiple objects to sign.

**Comment by tlk** [comment]
Have you checked out the `--no-quarantine` flag?

See https://github.com/Homebrew/homebrew-cask/blob/HEAD/doc/faq/app_cant_be_opened.md

**Reply by agologan** [reply]
As seen from our [README.md](https://github.com/liquidx/webviewscreensaver/blob/master/README.md) I already direct users to install with `--no-quarantine` to avoid the prompts on x86.
This does not answer my inquiry though, as codesigning is still required for Apple Sillicon as opposed to running on Intel.

**Reply by tlk** [reply]
Oh, I see. I guess you can maintain a local tap (as an alternative to core) with a formula.

Maybe reach out to Apple and ask them for advice? It would be interesting to know what their thoughts are on this subject.
